### ISSUE TITLE
nix::sys::socket::recv_from should accept MsgFlags

### ISSUE THREAD

**Issue by tgeoghegan** [nix-rust/nix#1203]
recvfrom(2) takes an integer `flags` parameter, but it is missing from [nix's version](https://docs.rs/nix/0.17.0/nix/sys/socket/fn.recvfrom.html):

`pub fn recvfrom(
    sockfd: RawFd, 
    buf: &mut [u8]
) -> Result<(usize, Option<SockAddr>)>`

The implementation of `recvfrom` unconditionally passes 0 to `libc::recvfrom`. This means I can't do something like `MSG_PEEK` without dropping down to `libc::recvfrom`. For context, [`nix::sys::socket::sendto`](https://docs.rs/nix/0.17.0/nix/sys/socket/fn.sendto.html) does correctly handle MsgFlags.

**Comment by asomers** [comment]
Can you use `recvmsg` with `None` for `cmsg_buffer` instead?  I think that should be equivalent.

**Comment by tgeoghegan** [comment]
I wound up using `libc::recvfrom` so I could set `MSG_PEEK` there. Fortunately, it operates on a `RawFd` so interop with nix level APIs is easy, and I already had to use `libc::socket` elsewhere (to use `SOCK_RAW`), so aside from forcing an `unsafe` block, it's not that bad of a workaround. 

The lack of `MsgFalgs` still seems like an omission from the nix level API though, especially since it lacks symmetry with `nix::sys::socket::sendto`. Having said that, I don't know how this could be fixed without breaking existing clients of `nix::sys::socket::recv_from`. Perhaps introduce a new `recv_from_with_flags`?

**Comment by asomers** [comment]
Nix is pre-1.0, so we do sometimes break backwards compatibility in small ways.  However, if `recvmsg` is an adequate workaround, which I suspect it is, then I'd rather just document that.

**Comment by tgeoghegan** [comment]
If I were in your shoes, and I knew I still had opportunities to fix API coming up, I'd fix this in an upcoming release where I knew I was going to break compatibility and force clients to update their usage anyway, rather than carry a wart like this forever. But otherwise I agree with your reluctance to inconvenience folks, and documenting a workaround seems like a reasonable fix.

**Comment by asomers** [comment]
I agree with the principle of not carrying warts forward.  However, I disagree with what the wart is.  Since `recvmsg` can do everything that `recvfrom` can, I think that the existence of `recvfrom` itself is the wart.
